### PR TITLE
Add Arm64 variants of build_all and test_all CI jobs

### PR DESCRIPTION
**Pull request by freddan80**
These jobs will run on the T2A runner.

### CODE REVIEW

**Comment by freddan80** [comment]
Related to #12684. 

I ran this locally using nektos/act, so hopefully this should work 'live' as well.

I tried using matrix strategy as well, which can eliminate some duplication. However, I didn't find a good way to handle output parameters, such as `build-dir` and `build-dir-archive`. They'll have to be duplicated with a matrix strategy in the `build_all.yml`, cause more and less readable code...

**Comment by freddan80** [comment]
Nah, it didn't work live. This work in my local setup :man_shrugging:  I'll look into it tomorrow.

```
[Invalid workflow file: .github/workflows/ci.yml#L70](https://github.com/openxla/iree/actions/runs/6576954474/workflow)
The workflow is not valid. In .github/workflows/ci.yml (Line: 70, Col: 11): Error from called workflow openxla/iree/.github/workflows/build_all.yml@41a151bf68cf0dc1ff0559e450c3aa62d94f6bc5 (Line: 54, Col: 18): Unrecognized named-value: 'env'. Located at position 2 within expression: (env.MACHINE_TYPE == 'arm64' && 'full-build-dir-arm64') ||
                                  'full-build-dir' openxla/iree/.github/workflows/build_all.yml@41a151bf68cf0dc1ff0559e450c3aa62d94f6bc5 (Line: 63, Col: 9): Unrecognized named-value: 'env'. Located at position 1 within expression: env.MACHINE_TYPE
```

File: .github/workflows/build_all.yml
Comment (pzread) [inline .github/workflows/build_all.yml]: I suspect this should be `${{ inputs.machine-type }}`. Github has inconsistent restrictions on where environment variables can be used.

Comment (freddan80) [reply .github/workflows/build_all.yml]: Yeah, I saw a similar discussion here:

https://github.com/actions/runner/issues/2372#issuecomment-1518528105

will fix

**Comment by freddan80** [comment]
Builds now, but very slow...

**Comment by dcaballe** [comment]
Really excited to see this moving forward!!!

File: .github/workflows/ci.yml
Comment (ScottTodd) [inline .github/workflows/ci.yml]: Please add new jobs to the `summary` job at the bottom of this file. That will make these jobs contribute to the overall build status: https://github.com/openxla/iree/blob/3a70ddae54bf62700d593e335c722464be1b4235/.github/workflows/ci.yml#L1082-L1101

File: .github/workflows/ci.yml
Comment (ScottTodd) [inline .github/workflows/ci.yml:1]: FYI, if you want to iterate on this PR using the CI, you can limit the jobs that run on this PR to just your new jobs by adding `ci-exactly: build_all_arm64, test_all_arm64` to your PR description (and the pushing a commit). See https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/contributing.md#ci-behavior-manipulation for more information about controlling the CI behavior.

**Comment by freddan80** [comment]
Veeery slow. Would it be possible to up the machine to "t2a-standard-16" (or even more)?

File: .github/workflows/ci.yml
Comment (ScottTodd) [inline .github/workflows/ci.yml]: > Builds now, but very slow...

We can test how much the caching helps with that. You can flip 'write-caches' to always enabled so presubmit runs populate the cache. By default, only postsubmit runs write to the cache (so presubmit runs on less trusted/stable code don't use storage space and "pollute" the remote cache).

Other things to check would be CPU core count and HDD/SSD setup on the machine.

**Comment by freddan80** [comment]
Or is it so that ccache need to "fill it's cache" ..?

File: .github/workflows/ci.yml
Comment (ScottTodd) [reply .github/workflows/ci.yml]: Keeping the discussion on a thread (wish GitHub handled this differently...)

> Veeery slow. Would it be possible to up the machine to "t2a-standard-16" (or even more)?

> Or is it so that ccache need to "fill it's cache" ..?

For builds of the compiler, we should use the largest machine we can find (within reason), _and_ use caching. Building LLVM on a 2 core machine can take hours, for example. We can configure the build a _bit_ to limit how much needs to be compiled (e.g. turning off LLVM targets like RISC-V and Wasm, turning off CUDA and Vulkan), but it's nice to get as much coverage as possible if we're already commiting to a long build.

File: .github/workflows/ci.yml
Comment (freddan80) [reply .github/workflows/ci.yml]: will do

File: .github/workflows/ci.yml
Comment (freddan80) [reply .github/workflows/ci.yml:1]: This is nice, thx

File: .github/workflows/ci.yml
Comment (freddan80) [reply .github/workflows/ci.yml]: You mean there should be two more runners, one x86 and one arm64, that handles the build_all jobs? Makes sense to me...

OTOH, the regular build job finished in 7m. Let's me try writing the cache.

Comment (freddan80) [reply .github/workflows/ci.yml]: Here's the first run with `write-cache` enabled:

https://github.com/openxla/iree/actions/runs/6584805240/job/17890046166?pr=15237

Then I re-ran:

https://github.com/openxla/iree/actions/runs/6586574344/job/17895211068?pr=15237

Unfortunately no significant change in build time. I must be missing something wrt ccache usage, this should be all in cache right?

We're using the `t2a-standard-8` runner which has 8 vCPU's ([T2A machines](https://cloud.google.com/compute/docs/general-purpose-machines#t2a_machines)). We can potentially up that to 48 vCPU's, but that still don't explain the ~5m compile time on `e2-medium` vs. ~70m on `t2a-standard-8`. I'm not sure how the `e2-medium` burst works too. Anyways, this also makes me believe the issue has to do with ccache, not the machine. The arm64 docker is using and older version of ccache, 3.7. (the Ubuntu 20.04 distro version), but I doubt that would make that big of a difference though... 

I do some digging in the logs and read up on ccache. Any ideas are most welcome :smile:

Comment (freddan80) [reply .github/workflows/ci.yml]: I see it's printed. So there's no cache hits at all from. See log from [here](https://github.com/openxla/iree/actions/runs/6586574344/job/17895211068?pr=15237):

```
stats updated                       Fri Oct 20 11:45:08 2023
cache hit (direct)                     0
cache hit (preprocessed)               0
cache miss                          5321
cache hit rate                      0.00 %
compile failed                         5
preprocessor error                    14
unsupported code directive             2
cleanups performed                   197
files in cache                      4930
cache size                           4.6 GB
max cache size                       5.0 GB
```

Trying to figure out why...

Comment (ScottTodd) [reply .github/workflows/ci.yml]: > You mean there should be two more runners, one x86 and one arm64, that handles the build_all jobs? Makes sense to me...

Not sure I follow here. I'm suggesting that we we should use the largest arm64 machine we can reasonably find when building the compiler. I don't know what the options there are - maybe these https://cloud.google.com/compute/docs/general-purpose-machines#t2a_machine_types ? No local SSD on any of those, but maybe "Zonal SSD (performance) Persistent Disk (`pd-ssd`)" is good?

Another option if we don't necessarily need to build and run the compiler on arm64 but still want to run the e2e tests on that platform would be to plug into the `cross_compile_and_test` job. That runs the compiler on a host (x86 Linux) platform and produces artifacts (.vmfb files produced by `iree-compile`) for using in e2e tests on target platforms like riscv and Android (Android is a little special and uses `build_and_test_android` though). Given the specs listed on the cloud page, we should be able to build the full compiler and not treat this as a cross compile job though.

---

RE ccache, if you have a development machine, debugging locally before with the remote cache could be easier. The remote setup looks correct (main thing is using a unique "CCACHE_NAMESPACE"). For local debugging, https://ccache.dev/manual/4.8.3.html#_cache_debugging has general info. I typically look at cache size, hit rate, cacheable calls, and then the debug logs to see if something in the configuration (choice of compiler, flags used, build platform) is unsupported or misconfigured.

Comment (pzread) [reply .github/workflows/ci.yml]: I chose `t2a-standard-8` to be more cost effective initially. We can change to `t2a-standard-48` if that is needed to achieve a reasonable build time. Our x86 build machine is `n1-standrd-96`, which is quite a powerful machine to compare with, but 70mins on the current `t2a-standard-8` seems to be abnormally long to me.

If you have T2A VM, what is the build time on your VM? I can also test the build on my side to see why it takes so long.

For ccache, IIUC it is expected to be 0 hit because we currently don't upload ccache from presubmit (this PR) to avoid cache poisoning. And there is no arm64 build existing in ccache before. So every build in this PR will be a new build. The ccache will be updated once it is built on the main branch.

Comment (pzread) [reply .github/workflows/ci.yml]: To clarify about the ccache, `write-cache` can be enabled in presubmit but presubmit runner won't have permissions to write the ccache storage currently.

We can test the build time with ccache on T2A VM locally to know the speedup we will see once this PR is merged. If locally it still takes too long to build, we will need to find other solutions (e.g. have a larger T2A runner).

Comment (freddan80) [reply .github/workflows/ci.yml]: > To clarify about the ccache, write-cache can be enabled in presubmit but presubmit runner won't have permissions to write the ccache storage currently.

Ok, I see. That explains why I got 0% cache hits. I looked for error messages related to writing remote ccache, so I assumed it worked.  

> Not sure I follow here. I'm suggesting that we we should use the largest arm64 machine we can reasonably find when building the compiler.

I thought [these runners](https://github.com/openxla/iree/blob/66d6968bb1b6dda3a98d651a6a758f6879af69f3/build_tools/github_actions/runner/gcp/create_image.sh#L40-L42) were the only runners currently used in this project. My question was whether you were suggesting to add another pair of "compile runners" to do the heavy lifting?

> We can test the build time with ccache on T2A VM locally to know the speedup we will see once this PR is merged. If locally it still takes too long to build, we will need to find other solutions (e.g. have a larger T2A runner).

That would be awesome. I suspect the compile time will be acceptable with ccache properly written.  

> If you have T2A VM, what is the build time on your VM? I can also test the build on my side to see why it takes so long.

> RE ccache, if you have a development machine, debugging locally before with the remote cache could be easier.

I use AWS machines, `c6g.16xlarge` and `c6i.16xlarge`. Without ccache, it takes about 15min. These are 64 vCPU machines, so with a crude translation to the  `t2a-standard-8` runner (8 vCPUs), the 75min compile time is reasonable. With ccache properly written, I expect/hope the compile time will drop to an acceptable level...

Comment (freddan80) [reply .github/workflows/ci.yml]: The [`test_all_arm64` job fails](https://github.com/openxla/iree/actions/runs/6586574344/job/17897279570#step:4:12) with this error when trying to fetch the `build_all` tarball:

```
Copying gs://iree-github-actions-presubmit-artifacts/6586574344/1/full-build-dir-arm64.tar.zst to file://full-build-dir-arm64.tar.zst
ERROR: [Errno 8] Exec format error: '/google-cloud-sdk/bin/gcloud-crc32c'
ERROR: Failed to download one or more component of sliced download.
...................................................
Average throughput: 957.3MiB/s
```

It probably because of the gcloud setup in [image_setup.sh](https://github.com/openxla/iree/blob/66d6968bb1b6dda3a98d651a6a758f6879af69f3/build_tools/github_actions/runner/gcp/image_setup.sh#L191-L197) is for x86. There seem to be no [corresponding file for arm64](https://cloud.google.com/sdk/docs/downloads-versioned-archives) to download. Is there a reason to curl Gcloud SDK as opposed to use apt, as described [here](https://cloud.google.com/sdk/docs/install#deb)? The specific version can be installed like this:

`sudo apt-get install google-cloud-cli=414.0.0-0`

However, I can validate it on the actual runner since I don't have access.

Comment (freddan80) [reply .github/workflows/ci.yml]: > With ccache properly written, I expect/hope the compile time will drop to an acceptable level...

I tried this locally and the cache miss rate goes down to ~45% from 100%. But it was still to slow unfortunately.

When upgrading to ccache 4.7.4, things look a lot better.  So I probably need to build ccache from source [here](https://github.com/openxla/iree/blob/446f96eb896dcc5a1d29ff0af3363d8df1fe7021/build_tools/docker/context/install_ccache.sh#L24-L26) instead, where I used the distro version. I'll make a PR.

Then I need your help to push the new image to the gcp registry. Then use that new docker...

Comment (pzread) [reply .github/workflows/ci.yml]: Thanks a lot for the testing! Sorry I'm a little bit occupied this week but I'll try to help ASAP

Comment (freddan80) [reply .github/workflows/ci.yml]: I created PR #15307. Let me know when you have time to push the new docker image to GCS. Then I'll update this PR with the new docker image.

Comment (freddan80) [reply .github/workflows/ci.yml]: I created #15308 to fix the issue mentioned [here](https://github.com/openxla/iree/pull/15237#discussion_r1368573696)

Comment (pzread) [reply .github/workflows/ci.yml]: Hi @freddan80 thanks for waiting. Once you merge #15307 I will push the docker image and post the SHA here.

Comment (pzread) [reply .github/workflows/ci.yml]: Oh, nvm, I can build image with your PR. Please replace
```
base-arm64@sha256:31cdb58f556e415c3c65afae31ae9c54e43c1eb3dcd845f3d107abd9b368ff3d
```
with
```
base-arm64@sha256:942d01a396f81bff06c5ef4643ba3c9f4500a09010cd30d9ed9be691ddaf1353
```

Comment (ScottTodd) [reply .github/workflows/ci.yml]: You should be able to propose those changes yourself, either by suggesting edits and accepting them or by pushing to the PR branch on the fork @pzread 
image

(I can also help here, LMK)

Comment (pzread) [reply .github/workflows/ci.yml]: Thanks for the tips. Done.

**Review by pzread — APPROVED** [review]
LGTM as long as the CI pass

**Comment by freddan80** [comment]
Once https://github.com/openxla/iree/pull/15308 is merged and the T2A runner is updated, I believe `test_all_arm64` should pass.

**Comment by pzread** [comment]
CI is passed now. Let me know if you need help to merge this change (and #15307). Or maybe @ScottTodd can provide up-to-date process to get write access to IREE repo (I remember it is to join https://github.com/orgs/openxla/teams/iree-write ?)

File: .github/workflows/ci.yml
Comment (ScottTodd) [reply .github/workflows/ci.yml]: How about we start these new jobs as postsubmit-only, while we monitor the build times and cache performance?

You can do that by adding the job name(s) to this list: https://github.com/openxla/iree/blob/fcdddcb8a79a30e5d32018ab3a47092f5e44534e/build_tools/github_actions/configure_ci.py#L118-L128

edit: (you can also opt-in to running the jobs on _presubmit_ for testing, see https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/contributing.md#ci-behavior-manipulation)

File: .github/workflows/ci.yml
Comment (ScottTodd) [inline .github/workflows/ci.yml]: As `test_all_arm64` is the only job that follows `build_all_arm64`, we could merge the jobs, like how `build_test_all_macos_arm64` is set up. Splitting the jobs is useful when multiple following jobs want to use the build artifacts, but otherwise it just adds overhead (upload/download, wait for runner, etc.).

File: .github/workflows/ci.yml
Comment (ScottTodd) [inline .github/workflows/ci.yml]: I think these should move down to the `# Platforms` group, next to `build_test_all_macos_arm64` etc.

**Review by ScottTodd — COMMENTED** [review]
Cool, this mostly LGTM. Just a few last comments. LMK if you want help with those or landing the change, happy to help get this over the finish line.

Thanks for your patience and sorry for all the back and forth. We also have some work planned to make updating the Docker images easier for collaborators too.

File: .github/workflows/ci.yml
Comment (freddan80) [reply .github/workflows/ci.yml]: I can do that. One thought though. If we'd like to extend with further arm64 server jobs in the future, isn't it convenient to keep them separated? (One example that has been discussed are the benchmark jobs.)

Comment (ScottTodd) [reply .github/workflows/ci.yml]: I don't think we'd want an arm64 server benchmark job to depend on `build_all_arm64`. It _could_, but it would introduce extra complexity.

Here's the current setup:
image

1. build on desktop x86
2. on desktop x86, generate benchmark files (for all architectures / targets)
    * this is one of the main features of a compiler like this - you can generate code for different architectures / platforms / targets from _any_ host machine
4. fork out to each target platform and execute the generated benchmarks
5. join back to a single machine to summarize the results and upload to the dashboard/database

On host platforms (linux x86 server, linux arm server, windows x86 server, macos arm server, etc.), we could go from `build_[platform]` -> `benchmark_[platform]`, but then we'd need to
* support each part of the infrastructure on all of those platforms
  * as you've been finding with gcloud CLI support and such, this takes a bit of effort
* maintain that the build jobs are consistently fast enough and enough runners are available to satisfy the CI load
  * power linux x86 server machines are relatively cheap compared to windows and macos server, and we don't yet fully know how fast the linux arm server builds will be

Hope that makes sense. I'm actually moving around / rewriting some docs now, and I see a good place in https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/contributing.md#ci for some more explanations and diagrams about this.

Comment (ScottTodd) [reply .github/workflows/ci.yml]: The benchmark jobs do need the `iree-benchmark-module` tool compiled for the target platform, which is handled by the `build_benchmark_tools` job. That has a matrix here: https://github.com/openxla/iree/blob/668c020cc04d66a60b31a6cda46ebdf3b854e952/.github/workflows/build_benchmark_tools.yml#L61-L79

Since that just builds runtime tools, it takes ~3 minutes.

Comment (freddan80) [reply .github/workflows/ci.yml]: Ok, so if I understood correctly the iree-compiler built as part of the `build_all` job is used to cross-compile everything that is going to be benchmark:ed?

Adding another target under [iree/.github/workflows/build_benchmark_tools.yml](https://github.com/openxla/iree/blob/668c020cc04d66a60b31a6cda46ebdf3b854e952/.github/workflows/build_benchmark_tools.yml#L61-L79) would create the executable needed to run that benchmark... About right?

Comment (ScottTodd) [reply .github/workflows/ci.yml]: Yep! @pzread is more familiar with `build_e2e_test_artifacts` than me though.

Comment (pzread) [reply .github/workflows/ci.yml]: Yes, that's right

Comment (freddan80) [reply .github/workflows/ci.yml]: Makes a lot of sense. Will do.

Comment (freddan80) [reply .github/workflows/ci.yml]: Done. I removed the changes in `build_all.yml` since they're not needed for this change.

Comment (freddan80) [reply .github/workflows/ci.yml]: Seems to pass. Let's see how what the build time drops to once the ccache is written.

File: .github/workflows/ci.yml
Comment (ScottTodd) [reply .github/workflows/ci.yml]: LGTM once the job is switched to postsubmit only (for now, while we monitor build stability and compile time)

Comment (freddan80) [reply .github/workflows/ci.yml]: Done

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**
